Support IPCC AR6 regions in `extract_shape`

schlunma opened the pull request:
<!--
    Thank you for contributing to our project!

    Please do not delete this text completely, but read the text below and keep
    items that seem relevant. If in doubt, just keep everything and add your
    own text at the top, a reviewer will update the checklist for you.

-->

## Description

This PR refactors the `extract_shape` preprocessor and makes it more flexible, by

- allowing a `dict` for the `ids` parameter to extract regions based on a non-default attribute name (e.g., `Acronym` for the IPCC AR6 regions). Default names are `id` and `name` (with different capitalization options).
- adding the shape file that describes the [IPCC AR6 regions](https://doi.org/10.5194/essd-12-2959-2020) to the repository and allowing easy access to that by using `shapefile: ar6`.

Example:

```yaml
extract_shape:
  shapefile: ar6
  method: contains
  ids:
    Acronym:
    - NEU
    - WCE
    - MED
```

<!--
    Please describe your changes here, especially focusing on why this pull
    request makes ESMValCore better and what problem it solves.

    Before you start, please read our contribution guidelines: https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html

    Please fill in the GitHub issue that is closed by this pull request,
    e.g. Closes #1903
-->

Closes #1648
Closes #758
Related to #2007

Link to documentation:

- https://esmvaltool--2008.org.readthedocs.build/projects/ESMValCore/en/2008/recipe/preprocessor.html#extract-shape
- https://esmvaltool--2008.org.readthedocs.build/projects/ESMValCore/en/2008/api/esmvalcore.preprocessor.html#esmvalcore.preprocessor.extract_shape

***

## [Before you get started](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#getting-started)

- [x] [☝ Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

<!-- The next two lines turn the 🛠 and 🧪 below into hyperlinks -->
[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [x] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [x] [🧪][2] and [🛠][1] [Documentation](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#documentation) is available
- [x] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [x] [🛠][1] Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)
- [x] [🛠][1] Any changed [dependencies have been added or removed](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#dependencies) correctly
- [x] [🛠][1] The [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors) is up to date
- [x] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful [2 Codacy issues about "Too many arguments (6/5) (too-many-arguments)" remain that will not be fixed].

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

valeriupredoi commented on esmvalcore/_recipe/recipe.py:
```diff
@@ -641,12 +642,32 @@ def _update_extract_shape(settings, session):
+                    shapefile.with_suffix('.shp')
+                )
+
+            settings['extract_shape']['shapefile'] = abs_shapefile
```
I would argue a DEBUG message would be nice to have per each of these looked into locations, that'd help the user see if they actually have some sort of shapefile, but it's not the one the code is looking for

schlunma replied:
Good idea, done in https://github.com/ESMValGroup/ESMValCore/pull/2008/commits/fb1dd77340bbf906089b06d445b53fff05c5268f

valeriupredoi replied:
cheers Manu! I'll get back to the rest of the review later today, a bit of a lax schedule for me these days with the Big C19 :grin:

valeriupredoi reviewed (commented):
quick comment through which I am signing myself up as a reviewer for this (cheers, Manu, very good idea!) but I got a meeting in 5min so I'll be back here after that :beer:

valeriupredoi commented on doc/recipe/preprocessor.rst:
```diff
@@ -1687,40 +1687,66 @@ See also :func:`esmvalcore.preprocessor.extract_named_regions`.
+      if no ``name`` nor ``id`` attributes are present, an integer ID starting
+      from 0 will be assigned automatically when reading the shapes.
+      We discourage to rely on this last behaviour as we can not assure that
+      the reading order will be the same in different platforms, so we
```
```suggestion
      the reading order will be the same on different platforms, so we
```

valeriupredoi commented on esmvalcore/preprocessor/_area.py:
```diff
@@ -440,108 +456,156 @@ def _correct_coords_from_shapefile(cube, cmor_coords, pad_north_pole,
+
+
+def _get_requested_geometries(
+    geometries: fiona.collection.Collection,
```
do we really want to hardassign types that may change in the future? ie fiona may change the name of `collection` without changing the actual API

schlunma replied:
Good point, will remove 👍

valeriupredoi commented on esmvalcore/preprocessor/_area.py:
```diff
@@ -440,108 +456,156 @@ def _correct_coords_from_shapefile(cube, cmor_coords, pad_north_pole,
+    if ids is not None:
+        missing = set(ids) - set(requested_geometries.keys())
+        if missing:
+            raise ValueError(f"Requested shapes {missing} not found")
```
plop a pointer to the shapefile too pls with `%s, shp_file`

valeriupredoi commented on esmvalcore/preprocessor/_area.py:
```diff
@@ -440,108 +456,156 @@ def _correct_coords_from_shapefile(cube, cmor_coords, pad_north_pole,
-    if method == 'representative' or not select.any():
-        select = _select_representative_point(shape, lon, lat)
-    return select
+        mask = shapely.vectorized.contains(shape, lon, lat)
```
this works in the new shapely >2 too right?

schlunma replied:
No idea, I guess we will see once the tests are run? Do we get shapely > 2 in our CI env now?

valeriupredoi replied:
I think so on Circle yes, which makes my question irrelevant :grin:

schlunma replied:
Tests ran fine again, so I guess we are safe here 🚀

valeriupredoi reviewed (approved):
this is very cool and very useful too - heard a few times about the need of AR6 regions in ESMValTool at the TerraFIRMA GA two weeks ago, very much cheers, Manu! I am approving because I don't want to be the ned stopping progress but I have a couple more serious questions (other than typos and minor code suggestions):

- how big are those shapefiles?
- I'd stay away from decalring types that may change in the future even if the external API may stay the same (eg fiona.collection.Collection)

Cheers, bud :beer:

schlunma commented on doc/recipe/preprocessor.rst:
```diff
@@ -1687,40 +1687,66 @@ See also :func:`esmvalcore.preprocessor.extract_named_regions`.
+    contained by the shape or a single representative point.
+    Choose either `'contains'` or `'representative'`.
+    If not a single grid point is contained in the shape, a representative
+    point will be selected.
```
Fully agree! Problem is that I don't understand that sentence myself 😅 My guess is that the nearest point in the data to the region is selected, but I am not sure. So I guess we just leave it like that?

valeriupredoi replied:
I don't either, so a rephrasing will not happen among the two of us, leave it for others to scratch their heads :rofl:

schlunma commented on doc/recipe/preprocessor.rst:
```diff
@@ -1687,40 +1687,66 @@ See also :func:`esmvalcore.preprocessor.extract_named_regions`.
+  * ``decomposed``: by default ``false``; in this case the union of all the
+    regions in the shapefile is masked out.
+    If ``true``, the regions in the shapefiles are masked out separately,
+    generating a new auxiliary dimension for the cube for this.
```
Good point, will change + add.

schlunma commented:
Thanks V, I addressed all your comments in https://github.com/ESMValGroup/ESMValCore/pull/2008/commits/8847b95f372d12330c777621e9097fbb629c6f88!

Regarding your questions:

- The total size of all AR6-related files is 380 K.
- Agreed about fiona type hints; I removed them 👍 

Thanks 🍻

valeriupredoi commented:
brilliant, you're a star, Manu! Ready to be merged then :partying_face:

bouweandela commented on esmvalcore/_recipe/recipe.py:
```diff
@@ -631,12 +632,44 @@ def _update_extract_shape(settings, session):
+
+            # If shapefile not found, first, try path relative to
+            # auxiliary_data_dir
+            if not abs_shapefile.exists():
+                abs_shapefile = session['auxiliary_data_dir'] / shapefile
+                logger.debug(
+                    "extract_shape: Looking for shapefile %s", abs_shapefile
+                )
+
+            # Second, try path relative to esmvalcore/preprocessor/shapefiles/
+            if not abs_shapefile.exists():
+                abs_shapefile = (
+                    Path(esmvalcore.preprocessor.__file__).parent /
+                    'shapefiles' /
+                    shapefile
+                )
+                logger.debug(
+                    "extract_shape: Looking for shapefile %s", abs_shapefile
```
The current implementation is only convenient for those who use a recipe instead of the preprocessor function directly. From a code maintenance and as well as from a user-friendliness perspective, it would nicer if this code was part of `_area.py` and users could just call the function with the argument `shapefile='AR6'` and that would look up the correct shapefile. To add a check here (or in check.py), you could just add the word 'AR6' as a valid shapefile name.

schlunma replied:
I agree that it would be nicer if one could use the argument `shapefile='ar6'` when calling `extract_shape` outside of running a recipe.

On the other hand, it's also nice to immediately get an error when running a recipe if the shapefile is invalid instead of getting the error after 1hr of preprocessing. I don't see a clear strategy here regarding this kind of problem when looking at other preprocessors...

Would it be an option to simply perform this check in `_recipe.py` **and** `_area.py`? I can put the code into `area.py` and then simply call it from from both locations.

valeriupredoi replied:
that's a very good point, Bouwe, and an even cooler point, Manu! We used to have file existence verifications in _recipe.py, back in the heady days of early 2.0 - I think there is a need for that, as well as a need for similar actions with various other files in auxialary_data_dir, so I fully support that!

schlunma replied:
Implemented in https://github.com/ESMValGroup/ESMValCore/pull/2008/commits/cdd5bebae75d4719cc4e33ff4fbc6823c78adf71

bouweandela replied:
Thanks! There appears to be an issue with building the documentation from the docstring of `extract_shape`, could you have a look please?

bouweandela commented on esmvalcore/_recipe/check.py:
```diff
@@ -238,7 +239,7 @@ def check_for_temporal_preprocs(profile):
     """Check that `extract_shape` arguments are valid."""
     shapefile = settings.get('shapefile', '')
-    if not os.path.exists(shapefile):
+    if not Path(shapefile).exists():
```
Note that Path(x) is a lot more fussy about what you put in than os.path.exists, so users may already get an exception from the error in creating a Path object from what they have defined instead of the intended RecipeError

bouweandela reviewed (commented):
Thanks @schlunma, great addition!

valeriupredoi commented:
then @bouweandela can merge, we found us a merger too :+1:
